pkg/agent: trigger batch before grpc message size overflow

heylongdacoder opened the pull request:
Fixes: #203

Signed-off-by: heylongdacoder <heylongdacoder@gmail.com>

heylongdacoder commented:
Hello, I don't have much experience on writing concurrent code, might make some mistake. Thanks for guidance in advance. :D

kakkoyun commented on pkg/agent/batch_remote_write_client.go:
```diff
@@ -32,21 +34,31 @@ type BatchWriteClient struct {
-	mtx    *sync.RWMutex
-	series []*profilestorepb.RawProfileSeries
+	triggerBatch chan struct{}
+	batchDone    chan struct{}
```
I think we can get rid of `batchDone` in here.

kakkoyun commented on pkg/agent/batch_remote_write_client.go:
```diff
@@ -67,45 +79,48 @@ func (b *BatchWriteClient) Run(ctx context.Context) error {
 	}
 }
 
 func (b *BatchWriteClient) batch(ctx context.Context) error {
```
```suggestion
func batch(ctx context.Contextm, r *profilestorepb.WriteRawRequest) error {
```

This method can receive the request and by this way we can get rid of the lock in this method if I'm not mistaken. This method could also omit the write client as a receiver if it's the case.

kakkoyun reviewed (changes requested):
Looking great!

I think we can simplify this code and make it more maintainable. We can get rid of one of the mutex locking/unlocking and remove one of the channels if we can make `batch` method to receive a request (probably we should rename it to just `send` after this change). With this flow will be easier and less error prone. 

What do you think?

heylongdacoder commented:
@kakkoyun thanks for the review and guidance! yes I need some help. I am not sure how would it works if we removed the `batchDone` channel. If we removed `batchDone`, I think there is a chance where `WriteRaw` function will continue to append data to the overflowed `writeRawRequest`(https://github.com/parca-dev/parca-agent/pull/836/files#diff-7276397f96515c11ca49f4f72de2237d63936ab985888a5a4fe07ccf8c371d42R166) before the `batch` function could send it. 

Currently I got another idea to avoid the creation of these `triggerBatch` and `batchDone` channels. Instead of a `writeRawRequest`, what about we create a slices of `writeRawRequest` and a `writeRawRequestIndex`? `WriteRaw` function will write to `writeRawRequest[writeRawRequestIndex]` and when the current one is going to be overflowed, then we write to `writeRawRequest[writeRawRequestIndex+1]`. Then the `batch` function either loop through the slices of `writeRawRequest` to fire gRPC request sequentially or create goroutine for each element to fire gRPC request. Do you think this would work and is this a better approach?

brancz commented:
Outdated since #2958 

We now use a totally new protocol that is apache-arrow based.
